### DISCUSSION TITLE
Unable to get proxy headers to work correctly in Keycloak 23.x

### DISCUSSION THREAD

**Discussion by dbs-dev** [keycloak/keycloak#26628] [Archive]
We have recently upgraded from a very old version of Keycloak, based on Wildfly, to Keycloak 23.0.5 and are struggling to get the remote IP addresses in the Keycloak events / logs. 

Our topology is like this: Web Application Firewall -> Kong -> Keycloak

We have enabled the reencrypt proxy mode and I confirm that in the Kong logs I can see that the following headers are set when forwarding the traffic to Keycloak:

```
"headers": {
     ...
    "x-proxy-remote-addr": "40.120.120.240",
    "x-real-ip": "40.120.120.240",
    "x-remote-addr": "40.120.120.240",
    "x-remote-ip": "40.120.120.240",
    "x-client-port": "5249",
    "x-forwarded-port": "443",
    "forwarded": "By=-.-;For=40.120.120.240;Host=dummy.fqdn.com;Proto=https",
    "x-forwarded-proto": "https",
    "x-forwarded_proto": "https",
    "x-forwarded-for": "40.120.120.240",
    "x-forwarded-server": "dummy.fqdn.com",
    "x-forwarded-host": "dummy.fqdn.com,
    "client-ip": "40.120.120.240",
    "x-originating-ip": "40.120.120.240"
    ...
}
```

I would expect the IP 40.120.120.240 to be in the logs, however, we see internal Kubernetes IP addresses.

What are we missing?

**Comment by vmuzikar** [comment]
@dbs-dev Sorry you're facing issues with the proxy config. Could you please share your output of the [hostname-debug tool](https://www.keycloak.org/server/hostname#_troubleshooting)?

**Comment by dbs-dev** [comment]
URL | Value
-- | --
Request | https://dummy.fqdn.com/auth/realms/master/hostname-debug
Frontend | https://dummy.fqdn.com/auth [OK]
Backend | https://dummy.fqdn.com/auth [OK]
Admin | https://dummy.fqdn.com/auth [OK]
Server mode | production [start]
Realm | master
hostname-url | https://dummy.fqdn.com/auth
hostname-admin-url | https://dummy.fqdn.com/auth
hostname-strict | true
hostname-strict-backchannel | true
hostname-strict-https | true
hostname-port | -1
proxy | reencrypt
http-enabled | false
http-relative-path | /auth
http-port | 8080
https-port | 8443
Forwarded | By=-.-;For=40.120.120.240;Host=dummy.fqdn.com;Proto=https
Host | dummy.fqdn.com
X-Forwarded-For | 40.120.120.240, 10.129.36.190
X-Forwarded-Host | dummy.fqdn.com
X-Forwarded-Port | 443
X-Forwarded-Proto | https

Note that I obfuscated some of the values. 
Could the issue be the fact that we see multiple entries in the X-Forwarded-For header?

**Reply by vmuzikar** [reply]
10.129.36.190 is the Kubernetes IP?

**Reply by dbs-dev** [reply]
Correct

**Reply by vmuzikar** [reply]
> Could the issue be the fact that we see multiple entries in the X-Forwarded-For header?

Doesn't seem like that. Your proxy sets both `Forwarded` and `X-Forwarded-` headers. Keycloak is transparently relying on Quarkus for parsing them. And it seems like Quarkus [prioritizes](https://github.com/quarkusio/quarkus/blob/3.2.9.Final/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ForwardedParser.java#L129) the `Forwarded` header.

Where exactly do you expect to see the client IP address (40.120.120.240) and the kube address is there instead?

**Reply by dbs-dev** [reply]
I expect to seem them in the event logs for the realms.

**Reply by vmuzikar** [reply]
This turned out to be actually a bug in Vert.x: https://github.com/quarkusio/quarkus/pull/38580

I'm not sure yet which version of Quarkus will this target but as soon as Keycloak upgrades to it, it'll be fixed implicitly for us too.

**Reply by dbs-dev** [reply]
Thanks for looking into this @vmuzikar! I will keep an eye on the Quarkus ticket.
